crates/cubecl-core/src/codegen/metadata.rs

```diff
@@ -0,0 +1,144 @@
+//! Metadata helpers to easily get offsets etc.
+//!
+//! Conceptually, metadata is represented like this:
+//! ```rust
+//! struct Metadata<const NUM_BUFS: usize, const NUM_EXT: usize> {
+//!     base: BaseMeta<NUM_BUFS>,
+//!     extended: ExtendedMeta<NUM_EXT>,
+//! }
+//!
+//! struct BaseMeta<const N: usize> {
+//!     buffer_lengths: [u32; N],
+//!     logical_lengths: [u32; N],
+//! }
+//!
+//! struct ExtendedMeta<const N: usize> {
+//!     ranks: [u32; N],
+//!     shape_offsets: [usize; N],
+//!     stride_offsets: [usize; N],
+//!     shapes: Vec<u32>,
+//!     strides: Vec<u32>
+//! }
+//! ```
+//! where `Vec` isn't an actual `Vec`, just a dynamically sized series of values.
+//!
+//! Ranks and lengths have a constant offset, while shapes/strides involve loading the tensor's
+//! offset, then adding `dim` to the offset to get each shape/stride.
+
+// Metadata
+const BUFFER_LEN: u32 = 0;
+const LENGTH: u32 = 1;
+const BASE_LEN: u32 = 2;
+
+// Extended Metadata
+const RANK: u32 = 0;
+const SHAPE_OFFSETS: u32 = 1;
+const STRIDE_OFFSETS: u32 = 2;
+
+#[derive(Clone, Debug, Default)]
+pub struct Metadata {
+    num_meta: u32,
+    num_extended_meta: u32,
+}
```

Comment:
Not sure what this struct does really 😅

Reply:
Forgot to add docs, but also it's just a helper that calculates the offset on the compiler side so we can change the layout without changing every usage in the backend.